File: examples3d/collision_groups3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Ground
     */
    let ground_size = 5.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(vector![0.0, -ground_height, 0.0]);
    let floor_handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
    colliders.insert_with_parent(collider, floor_handle, &mut bodies);

    /*
     * Setup groups
     */
    const GREEN_GROUP: InteractionGroups = InteractionGroups::new(Group::GROUP_1, Group::GROUP_1);
    const BLUE_GROUP: InteractionGroups = InteractionGroups::new(Group::GROUP_2, Group::GROUP_2);

    /*
     * A green floor that will collide with the GREEN group only.
     */
    let green_floor = ColliderBuilder::cuboid(1.0, 0.1, 1.0)
        .translation(vector![0.0, 1.0, 0.0])
        .collision_groups(GREEN_GROUP);
    let green_collider_handle =
        colliders.insert_with_parent(green_floor, floor_handle, &mut bodies);

    testbed.set_initial_collider_color(green_collider_handle, [0.0, 1.0, 0.0]);

    /*
     * A blue floor that will collide with the BLUE group only.
     */
    let blue_floor = ColliderBuilder::cuboid(1.0, 0.1, 1.0)
        .translation(vector![0.0, 2.0, 0.0])
        .collision_groups(BLUE_GROUP);
    let blue_collider_handle = colliders.insert_with_parent(blue_floor, floor_handle, &mut bodies);

    testbed.set_initial_collider_color(blue_collider_handle, [0.0, 0.0, 1.0]);

    /*
     * Create the cubes
     */
    let num = 8;
    let rad = 0.1;

    let shift = rad * 2.0;
    let centerx = shift * (num / 2) as f32;
    let centery = 2.5;
    let centerz = shift * (num / 2) as f32;

    for j in 0usize..4 {
        for i in 0..num {
            for k in 0usize..num {
                let x = i as f32 * shift - centerx;
                let y = j as f32 * shift + centery;
                let z = k as f32 * shift - centerz;

                // Alternate between the green and blue groups.
                let (group, color) = if k % 2 == 0 {
                    (GREEN_GROUP, [0.0, 1.0, 0.0])
                } else {
                    (BLUE_GROUP, [0.0, 0.0, 1.0])
                };

                let rigid_body = RigidBodyBuilder::dynamic().translation(vector![x, y, z]);
                let handle = bodies.insert(rigid_body);
                let collider = ColliderBuilder::cuboid(rad, rad, rad).collision_groups(group);
                colliders.insert_with_parent(collider, handle, &mut bodies);

                testbed.set_initial_body_color(handle, color);
            }
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point!(10.0, 10.0, 10.0), Point::origin());
}

File: bindings/python/rapier-testbed/rapier_testbed/examples3/collision_groups3.py
"""Port of ``examples3d/collision_groups3.rs``.

Two coloured floors and a stack of cubes split into ``GREEN`` and ``BLUE``
interaction groups. Cubes only land on the floor matching their group.
"""
from __future__ import annotations

import rapier3d as rp

from .._registry import register

CATEGORY = "Collisions"
NAME = "Collision groups"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 5.0
    ground_height = 0.1
    floor_body = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    floor_handle = bodies.insert(floor_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size),
        floor_handle,
        bodies,
    )

    green_group = rp.InteractionGroups(
        memberships=rp.Group.GROUP_1,
        filter=rp.Group.GROUP_1,
        test_mode=rp.InteractionTestMode.AND,
    )
    blue_group = rp.InteractionGroups(
        memberships=rp.Group.GROUP_2,
        filter=rp.Group.GROUP_2,
        test_mode=rp.InteractionTestMode.AND,
    )

    green_floor = (
        rp.Collider.cuboid(1.0, 0.1, 1.0)
        .translation((0.0, 1.0, 0.0))
        .collision_groups(green_group)
    )
    colliders.insert_with_parent(green_floor, floor_handle, bodies)

    blue_floor = (
        rp.Collider.cuboid(1.0, 0.1, 1.0)
        .translation((0.0, 2.0, 0.0))
        .collision_groups(blue_group)
    )
    colliders.insert_with_parent(blue_floor, floor_handle, bodies)

    num = 8
    rad = 0.1
    shift = rad * 2.0
    centerx = shift * (num // 2)
    centery = 2.5
    centerz = shift * (num // 2)

    for j in range(4):
        for i in range(num):
            for k in range(num):
                x = i * shift - centerx
                y = j * shift + centery
                z = k * shift - centerz

                group = green_group if k % 2 == 0 else blue_group
                body = rp.RigidBody.dynamic().translation((x, y, z))
                handle = bodies.insert(body)
                colliders.insert_with_parent(
                    rp.Collider.cuboid(rad, rad, rad).collision_groups(group),
                    handle,
                    bodies,
                )

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((10.0, 10.0, 10.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
